lib/src/main/scala/spinal/lib/Stream.scala

```diff
@@ -1403,45 +1410,143 @@ class StreamFifo[T <: Data](val dataType: HardType[T],
     }
   }
 
+  val asserts = globalData.config.formalAsserts generate formalAsserts()
+
+  def formalAsserts() = new Area {
```

Comment:
As I understand, from BMC to prove stage, there only required some assertion on internal states to synchronize the ramdomly initialized registers. If you could I suggest to add a small part assertion first. I think there are too many logics in this formalXXXX functions.

Btw, I am happy to see someone is using and contributing code to the formal related functions. I can also help, when I find enough time to reevaluate this part, later.

Reply:
I can refactor this into something more cohesive for sure. I think this is about the right amount of assertions to constrain the fifo though. 

My work flow for this is to run it, see where induction fails, patch the faulty state with an assert, and repeat as necessary. I'll also go through when cleaning up and comment out assertions and re-run it to see if they were necessary or not.